src/server/tiering/disk_storage.h

```diff
@@ -0,0 +1,46 @@
+// Copyright 2024, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#pragma once
+
+#include <string>
+#include <system_error>
+
+#include "core/external_alloc.h"
+#include "io/io.h"
+#include "server/io_mgr.h"
+
+namespace dfly::tiering {
+
+struct DiskSegment {
+  size_t offset, length;
+};
+
+// Disk storage controlled by asynchronous operations.
+class DiskStorage {
+ public:
+  using ReadCb = std::function<void(std::string_view)>;
+  using StashCb = std::function<void(DiskSegment)>;
+
+  std::error_code Open(std::string_view path);
+  void Close();
+
+  // Request read for segment, cb will be called on completion with read value
+  void Read(DiskSegment segment, ReadCb cb);
+
+  // Mark segment as free, performed immediately
+  void MarkAsFree(DiskSegment segment);
+
+  // Request bytes to be stored, cb will be called with assigned segment on completion. Can block to
+  // grow backing file, thus safe to call only from single fiber.
+  // Returns error code if operation failed  immediately (most likely it failed to grow the backing
+  // file) or passes an empty segment if the final write operation failed.
+  std::error_code Stash(io::Bytes bytes, StashCb cb);
```

Comment:
The interface essentially has two error reporting mechanisms, through error_code returns and through the callback. Maybe we should keep only one, maybe even make the callback accept io::result?

Reply:
depends if the caller needs to know whether the request was sent or not.
probably, can be changed later when we integrate it into tiered storage.

also relates to your comment about grow backing file - we can not assume a single fiber caller. in our main branch we allow to ignore stash requests, our high level logic allows it too. why do you want to change the contract?

Reply:
> we can not assume a single fiber caller

Isn't it currently managed by a single background job?

>  why do you want to change the contract?

Because it enforces a chaotic desing. Currently, if allocating space succeeded, we directly set io_pending = true on the prime value. For sure I don't want to couple DiskStorage with handling PrimeValues, so instead it should give an adquate response to the higher-level component that is in charge of the prime value. Of course we can ignore stash requests, but we should return an appropriate error

Reply:
I agree -  `the caller needs to know whether the request was sent or not.`

> Isn't it currently managed by a single background job?
Not necessarily. We also kick off the offloading directly from the request and we even have throttled writes 
`ScheduleOffloadWithThrottle` that is intended to offload most of the data during `debug populate` call to show off the speed of offloading.